### PR TITLE
Memory - Qdrant Vector Database Connector

### PR DESCRIPTION
**Pull request by tawalke**
### Motivation and Context

This pull request/PR review is to add the ability for the Semantic Kernel to persist embeddings/memory to external vector databases like Qdrant. This submission has modifications and additions to allow for integration into the current SK Memory architecture and the subsequent SDK/API of various Vector Databases. 

**Why is this change required?**
This change is required in order to allow SK developers/users to persist and search for embeddings/memory from a Vector Database like Qdrant.

**What problem does it solve?**
Adds capabilities of Long Term Memory/Embedding storage to a vector databases from the Semantic Kernel.

**What scenario does it contribute to?**
Scenario: Long term, scalability memory storage, retrieval, search, and filtering of Embeddings.

**If it fixes an open issue, please link to the issue here.**
N/A

### Description

This PR currently includes connection for the Qdrant VectorDB only, removing the initial Milvus VectorDB addition and VectorDB client interfaces for consistency across various external vector databases, which will be provided in forthcoming PR.

- Addition and Modification of Qdrant.Dotnet SDK
- Addition of new namespace Skills.Memory.QdrantDB
- Creating/Adding Qdrant Memory Client class and QdrantMemoryStore.cs. Adding methods for connecting, retrieving collections and embeddings from Qdrant vector database in cloud.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [X] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

**Comment by awharrison-28** [comment]
Please update the PR description with Motivation & Context as well as Changes made

File: samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            Console.WriteLine(collection);
+        }
+
+        Console.WriteLine("== Retrieving Memories ==");
+        MemoryQueryResult? lookup = await kernel.Memory.GetAsync(MemoryCollectionName, "cat1");
+        Console.WriteLine(lookup != null ? lookup.Metadata.Text : "No memories found");
+
+        Console.WriteLine("== Removing Collection {0} ==", MemoryCollectionName);
+        Console.WriteLine("== Printing Collections in DB ==");
```
Comment (lemillermicrosoft) [inline samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs:62]: nit:

```suggestion
        await memoryStore.DeleteCollectionAsync(MemoryCollectionName);
        Console.WriteLine("== Printing Collections in DB ==");
```

**Comment by awharrison-28** [comment]
Please update contribution checklist

File: samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs (hunk 1 above)
Comment (awharrison-28) [reply samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs]: addressed

File: dotnet/src/SemanticKernel/Memory/MemoryQueryResult.cs [hunk 2]
```diff
@@ -9,73 +12,55 @@ public class MemoryQueryResult
...
+    {
+        var metadata = JsonSerializer.Deserialize<MemoryRecordMetadata>(json);
+        if (metadata != null)
+        {
+            return new MemoryQueryResult(metadata, relevance);
+        }
+        else
+        {
+            throw new MemoryException(
+                MemoryException.ErrorCodes.UnableToSerializeMetadata,
```
[the remark this answers is not in the document]
Comment (awharrison-28) [reply dotnet/src/SemanticKernel/Memory/MemoryQueryResult.cs]: addressed

File: samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+// ReSharper disable once InconsistentNaming
+public static class Example19_Qdrant
+{
+    private const string MemoryCollectionName = "qdrant-test";
+
+    public static async Task RunAsync()
+    {
+        int qdrantPort = int.Parse(Env.Var("QDRANT_PORT"), CultureInfo.InvariantCulture);
+        QdrantMemoryStore memoryStore = new QdrantMemoryStore(Env.Var("QDRANT_ENDPOINT"), qdrantPort);
```
[the remark this answers is not in the document]
Comment (awharrison-28) [reply samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs]: addressed

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    <PropertyGroup>
+        <!-- NuGet Package Settings -->
+        <PackageId>Microsoft.SemanticKernel.Skills.Memory.Qdrant</PackageId>
+        <Title>Semantic Kernel - Qdrant Connector</Title>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Microsoft.CSharp" VersionOverride="4.7.0" />
+        <PackageReference Include="Microsoft.Extensions.Logging.Console" />
```
[the remark this answers is not in the document]
Comment (awharrison-28) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj]: removed the two packages from the comment

**Review by lemillermicrosoft — DISMISSED** [review]
Hooray getting this capability in!

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+using System.Net.Http;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.AI.Embeddings;
+using Microsoft.SemanticKernel.Memory.Storage;
+using Microsoft.SemanticKernel.Skills.Memory.Qdrant.Diagnostics;
+using Microsoft.SemanticKernel.Skills.Memory.Qdrant.Http;
+using Microsoft.SemanticKernel.Skills.Memory.Qdrant.Http.ApiSchema;
```
Comment (dluc) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:16]: this client should be equivalent to a client written by a third party, without any dependency on the kernel. All the mapping to what the kernel does with this should be in QdrantMemoryStore

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj [hunk 6]
```diff
@@ -0,0 +1,26 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <RepoRoot>$([System.IO.Path]::GetDirectoryName($([MSBuild]::GetPathOfFileAbove('.gitignore', '$(MSBuildThisFileDirectory)'))))</RepoRoot>
+    </PropertyGroup>
+    <Import Project="$(RepoRoot)/dotnet/nuget/nuget-package.props" />
+
+    <PropertyGroup>
+        <AssemblyName>Microsoft.SemanticKernel.Skills.Memory.Qdrant</AssemblyName>
```
Comment (dluc) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj:8]: I don't think this is a skill, this is a connector, and we should start having a namespace for connectors

**Review by dluc — CHANGES_REQUESTED** [review]

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+    {
+        this._log.LogDebug("Searching top {0} closest vectors in {1}", top);
+
+        Verify.NotNull(target, "The given vector is NULL");
+
+        using HttpRequestMessage request = SearchVectorsRequest<TEmbedding>
+            .Create(collectionName)
+            .SimilarTo(target.Vector.ToArray())
+            .HavingTags(requiredTags)
+            .IncludePayLoad()
```
Comment (awharrison-28) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:322]: SearchVectorsRequest is missing a score threshold

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+        {
+            yield return collection.Name;
+        }
+    }
+
+    #region private ================================================================================
+
+    private readonly ILogger _log;
+    private readonly HttpClient _httpClient;
+    private const int DefaultVectorSize = 1536; //output dimension size for OpenAI's text-emebdding-ada-002
```
Comment (awharrison-28) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:474]: This vector size should be probably passed in at object creation

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+        {
+            this._log.LogWarning("Vector not found");
+            return null;
+        }
+
+        var recordData = data.Result.First();
+
+#pragma warning disable CS8604 // The request specifically asked for a payload to be in the response
+        var record = new QdrantVectorRecord<TEmbedding>(
+            new Embedding<TEmbedding>(
```
Comment (awharrison-28) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:109]: Embedding should be replaced by float[] in QdrantVectorDbClient and its dependencies to @dluc's point above.

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj (hunk 6 above)
Comment (awharrison-28) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj:8]: I think we used to, and then we got rid of it, so there was some confusing on how to namespace this. Definitely agree that it is a connector.

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs (hunk 5 above)
Comment (awharrison-28) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:16]: @dluc will you be willing to accept this PR as is knowing that the next work I have in the pipeline is to merge the memory interfaces, make MemoryRecord the IMemoryStore return type, and eliminate generics from IMemoryStore? I can address the issue in this PR, but it's a pretty annoying, wide change that will be overwritten in the design we discussed earlier today.

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs (hunk 9 above)
Comment (awharrison-28) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:109]: See my other comment addressed to @dluc

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs (hunk 7 above)
Comment (awharrison-28) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:325]: done

File: samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            Console.WriteLine(collection);
+        }
+
+        Console.WriteLine("== Retrieving Memories ==");
+        MemoryQueryResult? lookup = await kernel.Memory.GetAsync(MemoryCollectionName, "cat1");
+        Console.WriteLine(lookup != null ? lookup.Metadata.Text : "ERROR: memory not found");
+
+        Console.WriteLine("== Similarity Searching Memories: My favorite color is orange ==");
+        var searchResults = kernel.Memory.SearchAsync(MemoryCollectionName, "My favorite color is orange", limit: 3, minRelevanceScore: 0.8);
```
Comment (lemillermicrosoft) [inline samples/dotnet/kernel-syntax-examples/Example19_Qdrant.cs:56]: Thanks for adding @awharrison-28

**Review by lemillermicrosoft — APPROVED** [review]
Agree with Devis re: Client not being tied to kernel. Would be okay addressing in a fast follow with other planned changes.

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantMemoryStore.cs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    public async Task<DataEntry<IEmbeddingWithMetadata<TEmbedding>>?> GetAsync(string collection, string key, CancellationToken cancel = default)
+    {
+        try
+        {
+            var vectorData = await this._qdrantClient.GetVectorByPayloadIdAsync(collection, key);
+            if (vectorData != null)
+            {
+                return new DataEntry<IEmbeddingWithMetadata<TEmbedding>>(
+                    key: key,
+                    value: vectorData?.Value);
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantMemoryStore.cs:46]:
```suggestion
value: vectorData.Value);
```

**Review by lemillermicrosoft — APPROVED** [review]

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs (hunk 8 above)
Comment (awharrison-28) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs]: done

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs (hunk 5 above)
Comment (tawalke) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/QdrantVectorDbClient.cs:16]: @dluc If what @awharrison-28 noted is okay I can resolve this conversation.

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj (hunk 6 above)
Comment (tawalke) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Skills.Memory.Qdrant.csproj:8]: @dluc I'm going to resolve this one and we can update w/future connectors.

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Diagnostics/IValidatable.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace Microsoft.SemanticKernel.Skills.Memory.Qdrant.Diagnostics;
+
+public interface IValidatable
```
Comment (dluc) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Diagnostics/IValidatable.cs:9]: does it need to be public?

Comment (tawalke) [reply dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/Diagnostics/IValidatable.cs:9]: I can verify and change it to internal

**Review by dluc — APPROVED** [review]
awesome work, thanks!

**Pull request merged**